File: cases/taint/cpp/mutual-recursive-transform-negative/mutual_recursive_transform_negative.cpp
int dfb_source() { // DFB-SOURCE: mutual-recursive-transform-input
    return 7;
}

void dfb_sink(int value) {} // DFB-SINK: mutual-recursive-transform-sink

int walk_b(int value, int depth);

int walk_a(int value, int depth) {
    if (depth == 0) {
        value = 0; // DFB-KILL: mutual-recursive-transform-a-base-clean
        return value;
    }
    int next = value + 1;
    int recursive = walk_b(next, depth - 1); // DFB-WITNESS: mutual-recursive-transform-a-transfer
    return recursive + 1; // DFB-WITNESS: mutual-recursive-transform-a-compose
}

int walk_b(int value, int depth) {
    if (depth == 0) {
        value = 0; // DFB-KILL: mutual-recursive-transform-b-base-clean
        return value; // DFB-WITNESS: mutual-recursive-transform-b-base
    }
    int next = value + 1;
    int recursive = walk_a(next, depth - 1); // DFB-WITNESS: mutual-recursive-transform-b-transfer
    return recursive + 1; // DFB-WITNESS: mutual-recursive-transform-b-compose
}

void run() {
    dfb_sink(walk_a(dfb_source(), 3));
}

File: cases/taint/scala/mutual-recursive-transform-negative/MutualRecursiveTransformNegative.scala
package dataflowbench

object MutualRecursiveTransformNegative {
  def dfb_source(): Int = { // DFB-SOURCE: mutual-recursive-transform-input
    7
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: mutual-recursive-transform-sink

  def walkA(value: Int, depth: Int): Int = {
    if (depth == 0) {
      var clean = value
      clean = 0 // DFB-KILL: mutual-recursive-transform-a-base-clean
      clean
    } else {
      val next = value + 1
      val recursive = walkB(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-a-transfer
      recursive + 1 // DFB-WITNESS: mutual-recursive-transform-a-compose
    }
  }

  def walkB(value: Int, depth: Int): Int = {
    if (depth == 0) {
      var clean = value
      clean = 0 // DFB-KILL: mutual-recursive-transform-b-base-clean
      clean // DFB-WITNESS: mutual-recursive-transform-b-base
    } else {
      val next = value + 1
      val recursive = walkA(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-b-transfer
      recursive + 1 // DFB-WITNESS: mutual-recursive-transform-b-compose
    }
  }

  def run(): Unit = {
    dfb_sink(walkA(dfb_source(), 3))
  }
}
